### PR TITLE
Add `break N` instead of `continue` and `continuing`.

### PR DESCRIPTION
**Pull request by jdashg**

### CODE REVIEW

File: wgsl/index.bs [hunk 1]
```diff
@@ -1058,52 +1053,95 @@ case_body
...
-* <1> the initializer is listed before the loop
-* <2> the continue construct is placed at the end of the `loop`.
 
-### Other Looping Constructs ### {#other-looping}
+### Other Proposed Looping Constructs ### {#other-looping}
 
 Note: This is *proposed* but not in the [SHORTNAME] spec yet
 
-The `for(var i : i32 = 0; i < 4; i = i + 1) {}` statement is syntactic sugar on
+The `for(var i : i32 = 0; i < 4; i = i + 1) {CODE}` statement is syntactic sugar on
```
Comment (kainino0x) [inline wgsl/index.bs]: Does this mean the `for`-loop no longer supports `continue` (i.e. you have to express it with a nested loop as with `loop`)?

Comment (jdashg) [reply wgsl/index.bs]: We could add it, but this is "proposed possibilities", as `for` isn't a real WGSL thing yet.

File: wgsl/index.bs [hunk 2]
```diff
@@ -1058,52 +1053,95 @@ case_body
...
+    var total_weight : i32 = 0;
     var i : i32 = 0; <1>
     loop {
-      break if (i >= 4);
-
-      a = a * 2;
-
-      continuing {  <2>
-        i = i + 1;
+      if (i < 4) break;
```
Comment (kainino0x) [inline wgsl/index.bs]:
```suggestion
      if (i >= 4) break;
```

File: wgsl/index.bs [hunk 3]
```diff
@@ -1058,52 +1053,95 @@ case_body
...
+      if (F) continue;
+      G
+      break;
+    }
+  </xmp>
+</div>
+
+<div class='example' heading="General [SHORTNAME] Loop Form">
+  <xmp>
+    A
```
Comment (kainino0x) [inline wgsl/index.bs]: Wrap this whole thing in `{}` to contain the scope of `A`?

**Comment by Kangz** [comment]
AFAIK the `break N` construct cannot be expressed in SPIR-V because of the [structured control flow rules](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_structuredcontrolflow_a_structured_control_flow) that says:

> the only blocks in a construct that can branch outside the construct are:
>
>  - ...
>  - a break block for the innermost loop it is nested inside of

. +CC @dj2 and @dneto0

**Comment by jdashg** [comment]
Hmm, suspicious. "a block branching to the construct’s merge block " satisfies some requirements, but I'll have to think it over.

File: wgsl/index.bs (hunk 3 above)
Comment (jdashg) [reply wgsl/index.bs]: Ah, yeah, that would be more precise.

**Comment by dneto0** [comment]
The literal count is error prone.  I'd rather label-name the loops, then have break _label_.

The translation from this more general construct requires, in general, synthesis of extra state variables to track mulit-level escape.  That adds complexity to translation, and is in conflict with the developer understanding (by counting) the register pressure on an invocation.

**Comment by kainino0x** [comment]
Removing approval  since it sounds like this can't translate to SPIR-V (directly).

**Comment by jdashg** [comment]
Ok, I agree that it would require non-zero-cost translation in the general case.

SPIR-V structured control flow branching rules:
> the only blocks in a construct that can branch outside the construct are
    * a block branching to the construct’s merge block
    * a block branching from one case construct to another, for the same OpSwitch
    * a back-edge block
    * a continue block for the innermost loop it is nested inside of
    * a break block for the innermost loop it is nested inside of
    * a return block

**Comment by jdashg** [comment]
In particular it would be temping to write:
```
let a = 0;
let y = 0;
loop {
  let x = 0;
  loop {
    a = foo(x,y);
    if (a) break 2;
    x += 1;
  }
  y += 1;
}
```

However, I believe there's no valid SPIR-V zero-cost equivalent.

**Pull request closed without merging**